algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -276,125 +274,120 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
                 circuit_term *= selector;
-                lincheck_sumcheck += (c.circuit_combiner, &circuit_term);
+
+                lineval_sumcheck += (c.circuit_combiner, &circuit_term);
             }
-            lincheck_sumcheck
-                .add(-v_H_at_beta, "h_1")
-                .add(-beta * g_1_at_beta, LCTerm::One);
-            lincheck_sumcheck
+            lineval_sumcheck
+                .add(-v_C_at_beta, "h_1")
+                .add(-beta * g_1_at_beta, LCTerm::One)
+                .add(-batch_lineval_sum, LCTerm::One);
+            lineval_sumcheck
         };
-        debug_assert!(evals.get_lc_eval(&lincheck_sumcheck, beta)?.is_zero());
+        debug_assert!(evals.get_lc_eval(&lineval_sumcheck, beta)?.is_zero());
 
-        for z_b in z_b_s.into_values() {
-            for z_b_i in z_b {
-                linear_combinations.insert(z_b_i.label.clone(), z_b_i);
-            }
-        }
         linear_combinations.insert("g_1".into(), g_1);
-        linear_combinations.insert("lincheck_sumcheck".into(), lincheck_sumcheck);
-        end_timer!(lincheck_time);
+        linear_combinations.insert("lineval_sumcheck".into(), lineval_sumcheck);
+        end_timer!(lineval_time);
 
         //  Matrix sumcheck:
         let mut matrix_sumcheck = LinearCombination::empty("matrix_sumcheck");
 
         for (i, (&id, circuit_state)) in state.circuit_specific_states.iter().enumerate() {
-            let v_H_i_at_alpha_time = start_timer!(|| format!("v_H_i_at_alpha {id}"));
-            let v_H_i_at_alpha = circuit_state.constraint_domain.evaluate_vanishing_polynomial(alpha);
-            end_timer!(v_H_i_at_alpha_time);
+            let v_R_i_at_alpha_time = start_timer!(|| format!("v_R_i_at_alpha {id}"));
+            let v_R_i_at_alpha = circuit_state.constraint_domain.evaluate_vanishing_polynomial(alpha);
+            end_timer!(v_R_i_at_alpha_time);
+
+            let v_C_i_at_beta_time = start_timer!(|| format!("v_C_i_at_beta {id}"));
+            let v_C_i_at_beta = circuit_state.variable_domain.evaluate_vanishing_polynomial(beta);
+            end_timer!(v_C_i_at_beta_time);
 
-            let v_H_i_at_beta_time = start_timer!(|| format!("v_H_i_at_beta {id}"));
-            let v_H_i_at_beta = circuit_state.constraint_domain.evaluate_vanishing_polynomial(beta);
-            end_timer!(v_H_i_at_beta_time);
-            let v_H_i = v_H_i_at_alpha * v_H_i_at_beta;
+            let v_rc = v_R_i_at_alpha * v_C_i_at_beta;
+            let RC = circuit_state.constraint_domain.size_as_field_element
+                * circuit_state.variable_domain.size_as_field_element;
 
             let s_a = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_a_domain, gamma);
             let s_b = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_b_domain, gamma);
             let s_c = selector_evals(&mut selectors, &max_non_zero_domain, &circuit_state.non_zero_c_domain, gamma);
```

Comment:
Ditto for this: probably worth precomputing the selector evaluations via a batch inversion method.

Reply:
See comment above